File: 2024/kotlin-2024/src/main/java/zone/frog/advent/twentyfour/Day1.kt
package zone.frog.advent.twentyfour

import java.io.File
import kotlin.math.abs

object Day1 {
    fun scenarioOne(textFile: String): Int {
        val left = ArrayList<Int>()
        val right = ArrayList<Int>()
        File(textFile).readLines()
            .forEach {
                with(it.split(" ")) {
                    left.add(first().toInt())
                    right.add(last().toInt())
                }
            }
        left.sort()
        right.sort()
        return left.zip(right)
            .map { abs(it.first - it.second) }
            .sum()
    }

    fun scenarioTwo(textFile: String): Int {
        val left = ArrayList<Int>()
        val right = ArrayList<Int>()
        File(textFile).readLines()
            .forEach {
                with(it.split(" ")) {
                    left.add(first().toInt())
                    right.add(last().toInt())
                }
            }
        return left.map { it * right.count { b -> it == b } }
            .sum()
    }
}

File: 2024/clojure/src/clojure_aoc/Day1.clj
(ns frog.zone)

(defn zip-lists [lists] (apply map vector lists))

(defn part-one [file]
  (->> (slurp file)
       (clojure.string/split-lines)
       (map #(->> (clojure.string/split % #"\s+")
                  (map read-string)
                  (vec)))
       (zip-lists)
       (map sort)
       (zip-lists)
       (map (fn [[a b]] (abs (- a b))))
       (reduce +)))

(part-one "../input/day1.txt")

(defn part-two [file]
  (let [[left right] (->> (slurp file)
                          (clojure.string/split-lines)
                          (map #(->> (clojure.string/split % #"\s+")
                                     (map read-string)
                                     (vec)))
                          (zip-lists))
        right-count (frequencies right)]
    (reduce + (map (fn [l] (* l (get right-count l 0))) left))))

(part-two "../input/day1.txt")
